### PR TITLE
Fully support aggregate on part of a input block

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: ref #7738

Problem Summary:

### What is changed and how it works?
This pr is after #8056, and in the pr, aggregate part of the input block in `Aggregator` fully supported. However, although supported, this feature can only be triggered by failpoint.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Operators/Operator.cpp [hunk 1]
```diff
@@ -162,7 +162,7 @@ OperatorStatus SinkOp::prepare()
     profile_info.anchor();
     auto op_status = prepareImpl();
 #ifndef NDEBUG
-    assertOperatorStatus(op_status, {OperatorStatus::NEED_INPUT});
+    assertOperatorStatus(op_status, {OperatorStatus::NEED_INPUT, OperatorStatus::IO_OUT});
```
Comment (SeaRise) [inline dbms/src/Operators/Operator.cpp:165]:
```suggestion
    assertOperatorStatus(op_status, {OperatorStatus::NEED_INPUT});
```
https://github.com/pingcap/tiflash/blob/c853837af6ecc170a854a19cc0a2d42cb12410aa/dbms/src/Operators/OperatorHelper.cpp#L29-L30

File: dbms/src/DataStreams/ParallelAggregatingBlockInputStream.cpp [hunk 2]
```diff
@@ -147,9 +147,12 @@ void ParallelAggregatingBlockInputStream::Handler::onBlock(Block & block, size_t
...
     agg_process_info.resetBlock(block);
-    parent.aggregator.executeOnBlock(agg_process_info, data, thread_num);
-    if (data.need_spill)
-        parent.aggregator.spill(data, thread_num);
+    do
+    {
+        parent.aggregator.executeOnBlock(agg_process_info, data, thread_num);
+        if (data.need_spill)
+            parent.aggregator.spill(data, thread_num);
+    } while (agg_process_info.start_row < agg_process_info.end_row);
```
Comment (SeaRise) [inline dbms/src/DataStreams/ParallelAggregatingBlockInputStream.cpp:155]:
```suggestion
    } while (!agg_process_info.isEnded());
```

File: dbms/src/DataStreams/ParallelAggregatingBlockInputStream.cpp [hunk 3]
```diff
@@ -273,6 +276,7 @@ void ParallelAggregatingBlockInputStream::execute()
         aggregator.executeOnBlock(agg_process_info, data, 0);
         if (data.need_spill)
             aggregator.spill(data, 0);
+        assert(agg_process_info.start_row == agg_process_info.end_row);
```
Comment (SeaRise) [inline dbms/src/DataStreams/ParallelAggregatingBlockInputStream.cpp:279]:
```suggestion
        assert(agg_process_info.isEnded());
```

**Review by yibin87 — APPROVED** [review]
LGTM

File: dbms/src/Operators/Operator.cpp (hunk 1 above)
Comment (windtalker) [reply dbms/src/Operators/Operator.cpp]: Done

File: dbms/src/Interpreters/Aggregator.cpp [hunk 4]
```diff
@@ -1059,17 +1067,26 @@ void Aggregator::execute(const BlockInputStreamPtr & stream, AggregatedDataVaria
...
-        if (is_cancelled())
-            return;
-
-        src_rows += block.rows();
-        src_bytes += block.bytes();
         agg_process_info.resetBlock(block);
+        bool should_stop = false;
+        do
+        {
+            if (is_cancelled())
```
Comment (SeaRise) [inline dbms/src/Interpreters/Aggregator.cpp:1074]:
```suggestion
            if unlikely (is_cancelled())
```

**Review by SeaRise — APPROVED** [review]
lgtm

**Pull request merged**